File: src/main/java/frc/robot/Robot.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot;

import java.util.ArrayList;

import edu.wpi.first.apriltag.AprilTag;
import edu.wpi.first.apriltag.AprilTagFieldLayout;
import edu.wpi.first.apriltag.AprilTagFields;
import edu.wpi.first.math.MathUtil;
import edu.wpi.first.math.filter.SlewRateLimiter;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Pose3d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Rotation3d;
import edu.wpi.first.math.geometry.Transform2d;
import edu.wpi.first.math.geometry.Transform3d;
import edu.wpi.first.networktables.NetworkTableInstance;
import edu.wpi.first.networktables.StructArrayPublisher;
import edu.wpi.first.networktables.StructPublisher;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.Joystick;
import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj.DriverStation.Alliance;
import edu.wpi.first.wpilibj.smartdashboard.Field2d;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.CommandScheduler;

public class Robot extends TimedRobot {
  private final Joystick m_controller = new Joystick(0);
  private final Drivetrain m_swerve = new Drivetrain();

  // Slew rate limiters to make joystick inputs more gentle; 1/3 sec from 0 to 1.
  private final SlewRateLimiter m_xspeedLimiter = new SlewRateLimiter(3);
  private final SlewRateLimiter m_yspeedLimiter = new SlewRateLimiter(3);
  private final SlewRateLimiter m_rotLimiter = new SlewRateLimiter(3);

  // Field to display robot location
  private final Field2d m_field = new Field2d();

  // Key poses
  private final Pose2d m_goalWRTField = new Pose2d(16.54, 5.55, Rotation2d.fromDegrees(180));
  private final Transform3d m_botToAngledCamera = new Transform3d(0, 0, .5, new Rotation3d(0, -Math.PI/4, 0));
  private final Transform3d m_botToFrontCamera = new Transform3d(.5, 0, .5, new Rotation3d(0, 0, 0));

  private final AprilTagFieldLayout fieldLayout;

  StructPublisher<Pose2d> botWRTAlliancePublisher = NetworkTableInstance.getDefault()
      .getStructTopic("/SmartDashboard/botWRTAlliance", Pose2d.struct).publish();
  StructArrayPublisher<Pose3d> tag3dPublisher = NetworkTableInstance.getDefault()
      .getStructArrayTopic("/SmartDashboard/3D/AprilTags", Pose3d.struct).publish();
  StructPublisher<Transform2d> robotToGoalPublisher = NetworkTableInstance.getDefault()
      .getStructTopic("/SmartDashboard/2D/robotToGoal", Transform2d.struct).publish();
  StructPublisher<Transform2d> goalToRobotPublisher = NetworkTableInstance.getDefault()
      .getStructTopic("/SmartDashboard/2D/goalToRobot", Transform2d.struct).publish();
  StructPublisher<Transform2d> robotToTag1Publisher = NetworkTableInstance.getDefault()
      .getStructTopic("/SmartDashboard/2D/robotToTag1", Transform2d.struct).publish();
  StructPublisher<Transform2d> tag1ToRobotPublisher = NetworkTableInstance.getDefault()
      .getStructTopic("/SmartDashboard/2D/tag1ToRobot", Transform2d.struct).publish();
  StructPublisher<Transform3d> angledCameraToTag1Publisher = NetworkTableInstance.getDefault()
      .getStructTopic("/SmartDashboard/3D/angledCameraToTag1", Transform3d.struct).publish();
  StructPublisher<Transform3d> frontCameraToTag1Publisher = NetworkTableInstance.getDefault()
      .getStructTopic("/SmartDashboard/3D/frontCameraToTag1", Transform3d.struct).publish();
  StructPublisher<Pose3d> angledCameraWRTAlliancePublisher = NetworkTableInstance.getDefault()
      .getStructTopic("/SmartDashboard/3D/angledCameraWRTAlliance", Pose3d.struct).publish();
      StructPublisher<Pose3d> frontCameraWRTAlliancePublisher = NetworkTableInstance.getDefault()
      .getStructTopic("/SmartDashboard/3D/frontCameraWRTAlliance", Pose3d.struct).publish();

  public Robot() {
    // Load the April Tag field layout from the resource file
    try {
      fieldLayout = AprilTagFieldLayout.loadFromResource(AprilTagFields.k2024Crescendo.m_resourceFile);
    } catch (Exception e) {
      System.err.println(e);
      throw new RuntimeException(e);
    }
  }

  @Override
  public void robotInit() {
    SmartDashboard.putData("Field", m_field);

  }

  @Override
  public void robotPeriodic() {
    CommandScheduler.getInstance().run();

    // Prepare field / alliance conversion for later
    Pose2d fieldWRTAlliance = fieldWRTAlliance();
    Pose2d allianceWRTField = allianceWRTField();
    Pose3d allianceWRTField3d = pose2dToPose3d(allianceWRTField);

    // Display robot location
    Pose2d botWRTAlliance = m_swerve.getRobotPose();    // << Look here
    Pose2d botWRTField = botWRTAlliance.relativeTo(fieldWRTAlliance);
    botWRTAlliancePublisher.set(botWRTAlliance);
    m_field.setRobotPose(botWRTAlliance);

    // Get Tag 1 pose
    Pose2d tag1WRTField = fieldLayout.getTags().get(0).pose.toPose2d();

    // Get important transforms
    Transform2d robotToGoal = m_goalWRTField.minus(botWRTField);
    Transform2d robotToTag1 = tag1WRTField.minus(botWRTField);
    Transform2d tag1ToRobot = robotToTag1.inverse();        // << Look here
    Transform2d goalToRobot = robotToGoal.inverse();

    // Publish transforms
    robotToGoalPublisher.set(robotToGoal);
    goalToRobotPublisher.set(goalToRobot);
    robotToTag1Publisher.set(robotToTag1);
    tag1ToRobotPublisher.set(tag1ToRobot);


    // Push April Tag positions to Smart Dashboard
    // AdvantageScope expects things in the alliance frame so we need to do some conversion. 
    var poseList = new ArrayList<Pose3d>();
    for (AprilTag tag : fieldLayout.getTags()) {
      var tagWRTField = tag.pose;
      var tagWRTAlliance = tagWRTField.relativeTo(allianceWRTField3d);
      poseList.add(tagWRTAlliance);
    }
    tag3dPublisher.set(poseList.toArray(new Pose3d[0]));

    // Add field objects to display in the odometry tab
    // AdvantageScope expects things in the alliance frame so we need to do some conversion.
    var goalWRTAlliance = m_goalWRTField.relativeTo(allianceWRTField);
    m_field.getObject("Goal").setPose(goalWRTAlliance);
    for (AprilTag tag : fieldLayout.getTags()) {
      var tagWRTField = tag.pose;
      var tagWRTAlliance = tagWRTField.relativeTo(allianceWRTField3d);
      m_field.getObject("Tag " + tag.ID).setPose(tagWRTAlliance.toPose2d());
    }


    // Get the 3D pose of the robot
    Pose3d botWRTField3d = pose2dToPose3d(botWRTField);
    Pose3d botWRTAlliance3d = pose2dToPose3d(botWRTAlliance);

    // Get the 3D pose of tag 1
    Pose3d tag1WRTField3d = fieldLayout.getTags().get(0).pose;

    // Get the 3D pose of the camera
    Pose3d angledCameraWRTAlliance = botWRTAlliance3d.transformBy(m_botToAngledCamera);
    Pose3d angledCameraWRTField = botWRTField3d.transformBy(m_botToAngledCamera);       // << Look here
    Pose3d frontCameraWRTAlliance = botWRTAlliance3d.transformBy(m_botToFrontCamera);
    Pose3d frontCameraWRTField = botWRTField3d.transformBy(m_botToFrontCamera);

    // Get the interesting transforms
    Transform3d angledCameraToTag1 = tag1WRTField3d.minus(angledCameraWRTField);
    Transform3d frontCameratoTag1 = tag1WRTField3d.minus(frontCameraWRTField);

    // Publish to Smart Dashboard
    angledCameraToTag1Publisher.set(angledCameraToTag1);
    frontCameraToTag1Publisher.set(frontCameratoTag1);
    angledCameraWRTAlliancePublisher.set(angledCameraWRTAlliance);
    frontCameraWRTAlliancePublisher.set(frontCameraWRTAlliance);
  }

  @Override
  public void autonomousPeriodic() {
    driveWithJoystick(false);
  }

  @Override
  public void teleopPeriodic() {
    driveWithJoystick(true);
  }

  private static Pose2d fieldWRTAlliance() {
    var alliance = DriverStation.getAlliance();
    if (!alliance.isEmpty() && alliance.get() == Alliance.Red)
      return new Pose2d(16.54, 8.21, Rotation2d.fromDegrees(180));
    else
      return new Pose2d();
  }

  private static Pose2d allianceWRTField() {
    // This is a weird special case where fiel to alliance is the same as alliance to field
    return fieldWRTAlliance();
  }

  private void driveWithJoystick(boolean fieldRelative) {
    // Get chassis speeds from joystick inputs.
    final var xSpeed = -m_xspeedLimiter.calculate(MathUtil.applyDeadband(m_controller.getRawAxis(1), 0.02))
        * Drivetrain.kMaxSpeed;

    final var ySpeed = -m_yspeedLimiter.calculate(MathUtil.applyDeadband(m_controller.getRawAxis(0), 0.02))
        * Drivetrain.kMaxSpeed;

    final var rot = -m_rotLimiter.calculate(MathUtil.applyDeadband(m_controller.getRawAxis(2), 0.02))
        * Drivetrain.kMaxAngularSpeed;

    // Move the robot with joystick inputs
    m_swerve.drive(xSpeed, ySpeed, rot, fieldRelative);
  }

  private static Pose3d pose2dToPose3d(Pose2d pose2d) {
    return new Pose3d(pose2d.getX(), pose2d.getY(), 0,
        new Rotation3d(0, 0, pose2d.getRotation().getRadians()));
  }
}

File: robot.py

#!/usr/bin/env python3
#
# Copyright (c) FIRST and other WPILib contributors.
# Open Source Software you can modify and/or share it under the terms of
# the WPILib BSD license file in the root directory of this project.
#

from math import pi
from robotpy_apriltag import AprilTagFieldLayout, AprilTagField
import wpilib
import wpimath
import wpilib.drive
import wpimath.filter
import wpimath.controller
import drivetrain
import ntcore
from wpimath.geometry._geometry import Pose2d, Pose3d, Transform2d, Transform3d, Rotation2d, Rotation3d

# Key poses
algaeWRTAlliance = Pose2d(1.2, 4.04, Rotation2d())
botToAngledCamera = Transform3d(0, 0, .5, Rotation3d(0, -pi/4, 0))
botToFrontCamera = Transform3d(.5, 0, .5, Rotation3d(0, 0, 0))

fieldLayout = AprilTagFieldLayout.loadField(AprilTagField.kDefaultField)
numTags = len(fieldLayout.getTags())


botWRTAlliancePublisher = ntcore.NetworkTableInstance.getDefault().getStructTopic("/SmartDashboard/botWRTAlliance", Pose2d).publish()
tag3dPublisher = ntcore.NetworkTableInstance.getDefault().getStructArrayTopic("/SmartDashboard/3D/AprilTags", Pose3d).publish()
robotToalgaePublisher = ntcore.NetworkTableInstance.getDefault().getStructTopic("/SmartDashboard/2D/robotToalgae", Transform2d).publish()
algaeToRobotPublisher = ntcore.NetworkTableInstance.getDefault().getStructTopic("/SmartDashboard/2D/algaeToRobot", Transform2d).publish()
robotToTag1Publisher = ntcore.NetworkTableInstance.getDefault().getStructTopic("/SmartDashboard/2D/robotToTag1", Transform2d).publish()
tag1ToRobotPublisher = ntcore.NetworkTableInstance.getDefault().getStructTopic("/SmartDashboard/2D/tag1ToRobot", Transform2d).publish()
tag1WRTAlliancePublisher = ntcore.NetworkTableInstance.getDefault().getStructTopic("/SmartDashboard/2D/tag1WRTAlliance", Pose2d).publish()
algaeWRTAlliancePublisher = ntcore.NetworkTableInstance.getDefault().getStructTopic("/SmartDashboard/2D/algaeWRTAlliance", Pose2d).publish()
angledCameraToTag1Publisher = ntcore.NetworkTableInstance.getDefault().getStructTopic("/SmartDashboard/3D/angledCameraToTag1", Transform3d).publish()
frontCameraToTag1Publisher = ntcore.NetworkTableInstance.getDefault().getStructTopic("/SmartDashboard/3D/frontCameraToTag1", Transform3d).publish()
angledCameraWRTAlliancePublisher = ntcore.NetworkTableInstance.getDefault().getStructTopic("/SmartDashboard/3D/angledCameraWRTAlliance", Pose3d).publish()
frontCameraWRTAlliancePublisher = ntcore.NetworkTableInstance.getDefault().getStructTopic("/SmartDashboard/3D/frontCameraWRTAlliance", Pose3d).publish()


aprilTagRedOrigin = Pose3d(1.01, -.15, 0, Rotation3d())
aprilTagBlueOrigin = Pose3d(18.558, 7.902, 0, Rotation3d(0, 0, 180))

def pose2dToPose3d(pose2d: Pose2d) -> Pose3d:
    return Pose3d(pose2d.x, pose2d.y, 0, Rotation3d(0, 0, pose2d.rotation().radians()))
  


class MyRobot(wpilib.TimedRobot):
    def robotInit(self) -> None:
        """Robot initialization function"""
        self.controller = wpilib.XboxController(0)
        self.swerve = drivetrain.Drivetrain()

        # Slew rate limiters to make joystick inputs more gentle 1/3 sec from 0 to 1.
        self.xspeedLimiter = wpimath.filter.SlewRateLimiter(3)
        self.yspeedLimiter = wpimath.filter.SlewRateLimiter(3)
        self.deadband = 0.1
        self.rotLimiter = wpimath.filter.SlewRateLimiter(3)

    def autonomousPeriodic(self) -> None:
        if self.isSimulation():
            self.simulationPeriodic()
                  
        self.driveWithJoystick(False)
        


    def teleopPeriodic(self) -> None:
        if self.isSimulation():
            self.simulationPeriodic()
                        
        self.driveWithJoystick(True)
        self.swerve.updateOdometry()

        fieldLayout.setOrigin(fieldLayout.OriginPosition.kBlueAllianceWallRightSide if wpilib.DriverStation.getAlliance() == wpilib.DriverStation.Alliance.kBlue else fieldLayout.OriginPosition.kRedAllianceWallRightSide)





        # Display robot location
        botWRTAlliance = self.swerve.odometry.getPose()   # << Look here
        botWRTAlliancePublisher.set(botWRTAlliance)





        # Get Tag 1 pose
        tag1WRTAlliance = fieldLayout.getTagPose(1).toPose2d()

        # Get important transforms
        robotToalgae = algaeWRTAlliance - botWRTAlliance
        robotToTag1 = tag1WRTAlliance - botWRTAlliance
        tag1ToRobot = robotToTag1.inverse()        # << Look here
        algaeToRobot = robotToalgae.inverse()

        # Publish transforms
        robotToalgaePublisher.set(robotToalgae)
        algaeToRobotPublisher.set(algaeToRobot)
        robotToTag1Publisher.set(robotToTag1)
        tag1ToRobotPublisher.set(tag1ToRobot)
        tag1WRTAlliancePublisher.set(tag1WRTAlliance)
        algaeWRTAlliancePublisher.set(algaeWRTAlliance)





        # Push April Tag positions to Smart Dashboard
        # AdvantageScope expects things in the alliance frame so we need to do some conversion. 
        poseList = []
        for tagNum in range(1, numTags+1):
            tagWRTAlliance = fieldLayout.getTagPose(tagNum)
            poseList.append(tagWRTAlliance)
        
        tag3dPublisher.set(poseList)

        # Get the 3D pose of the robot
        botWRTAlliance3d = pose2dToPose3d(botWRTAlliance)

        # Get the 3D pose of tag 1
        tag1WRTAlliance3d = fieldLayout.getTagPose(1)

        # Get the 3D pose of the camera
        angledCameraWRTAlliance = botWRTAlliance3d + botToAngledCamera   # << Look here
        frontCameraWRTAlliance = botWRTAlliance3d + botToFrontCamera

        # Get the interesting transforms
        angledCameraToTag1 = tag1WRTAlliance3d - angledCameraWRTAlliance
        frontCameratoTag1 = tag1WRTAlliance3d - frontCameraWRTAlliance

        # Publish to Smart Dashboard
        angledCameraToTag1Publisher.set(angledCameraToTag1)
        frontCameraToTag1Publisher.set(frontCameratoTag1)
        angledCameraWRTAlliancePublisher.set(angledCameraWRTAlliance)
        frontCameraWRTAlliancePublisher.set(frontCameraWRTAlliance)
        botWRTAlliancePublisher.set(self.swerve.odometry.getPose())
        

    def simulationPeriodic(self) -> None:
        self.swerve.updateSimulation(self.getPeriod())
		

    def driveWithJoystick(self, fieldRelative: bool) -> None:
        # Get the x speed. We are inverting this because Xbox controllers return
        # negative values when we push forward.
        xSpeed = (
            -self.xspeedLimiter.calculate(
                wpimath.applyDeadband(self.controller.getLeftY(), self.deadband)
            )
            * drivetrain.kMaxSpeed
        )

        # Get the y speed or sideways/strafe speed. We are inverting this because
        # we want a positive value when we pull to the left. Xbox controllers
        # return positive values when you pull to the right by default.
        ySpeed = (
            -self.yspeedLimiter.calculate(
                wpimath.applyDeadband(self.controller.getLeftX(), self.deadband)
            )
            * drivetrain.kMaxSpeed
        )

        # Get the rate of angular rotation. We are inverting this because we want a
        # positive value when we pull to the left (remember, CCW is positive in
        # mathematics). Xbox controllers return positive values when you pull to
        # the right by default.
        rot = (
            -self.rotLimiter.calculate(
                wpimath.applyDeadband(self.controller.getRightX(), self.deadband)
            )
            * drivetrain.kMaxSpeed
        )


        self.swerve.drive(xSpeed, ySpeed, rot, fieldRelative, self.getPeriod())

    